### DISCUSSION TITLE
Discovery Behaviour

### DISCUSSION THREAD

**Discussion by Iamknownasfesal** [libp2p/rust-libp2p#3864] [Q&A] [answered]
Hey, we are trying to implement DHT, Gossipsub, Heartbeats, Request & Response.
We are having issue in discovery behaviour, which is essential for our blockchain application.

**We have 100 nodes in a network, which they need to discover nodes via bootstrap nodes, and create rounds.

But seems like limiting discover numbers limits the discover itself. Any new node cannot connect to bootstrap node, or cannot discover any nodes due to losing connection to the bootstrap, because of discover limit.**

How can we solve this issue?

**Comment by thomaseizinger** [comment]
We don't have a dedicated "discovery behaviour", can you elaborate what you are referring to?

Also, by default, there are no connection limits. Are you setting them yourself?

**Reply by Iamknownasfesal** [reply]
As behaviour, i mean that i built a behaviour with DHT(kademlia) etc etc. I have a discovery limit of 50, which i want one node to connect only 50 nodes. 

But in the same way, new connections should be directed to the empty nodes, which they will have open slots for connections.

What happens if i not select any limit in Connection Limit on Swarm, does libp2p accept all connections until ports ends?

**Reply by Iamknownasfesal** [reply]
![image](https://user-images.githubusercontent.com/39006465/235695644-75ca4e71-414a-4108-8c3e-f672a19c278f.png)

**Reply by thomaseizinger** [reply, marked as answer]
Two things:

- In the latest `libp2p` version, `connection_limits` is deprecated, you should use the `libp2p::connections_limits::Behaviour` instead.
- You kind of have to decide what you want:
  With a counter-based connection limit, we will disallow new connections if they exceed it. We cannot know ahead of time what the "purpose" of this connection is, i.e. whether it is an initial "bootstrap" connection or not. If you need a more flexible connection management, you are probably better of with implementing `NetworkBehaviour` yourself. Take a look at [`libp2p::connection_limits::Behaviour`](https://github.com/libp2p/rust-libp2p/blob/master/misc/connection-limits/src/lib.rs) for some inspiration.

> What happens if i not select any limit in Connection Limit on Swarm, does libp2p accept all connections until ports ends?

Yes.

> But in the same way, new connections should be directed to the empty nodes, which they will have open slots for connections.

There are several ways to achieve this. If new nodes only know about the bootstrap nodes, then what you need is for them to initially connect to the bootstrap nodes, discover new peers via the DHT and then **disconnect** from the bootstrap nodes, freeing up slots for new connections. You can write yourself your own `NetworkBehaviour` that records and persists addresses of nodes you have been connected to such that on the next startup, you can connect to the other nodes directly and don't have to go through the bootstrap nodes again.

**Reply by Iamknownasfesal** [reply]
Thank you thomas, i will implement the new library as you said. 

And also will implement custom network behaviour, seems valid to me. Thank you!!
